drivers/char/rust_example/src/lib.rs

```diff
@@ -36,8 +51,14 @@ impl KernelModule for RustExample {
         println!("Parameters:");
         println!("  my_bool:  {}", my_bool.read());
         println!("  my_i32:   {}", my_i32.read());
+
+        let mut d = kernel::try_alloc_pinned(miscdev::Registration::new())?;
```

Comment:
Is there a reason not to have `Registration::new()` do the pin? is there ever a use case for a non-pinned value?

Reply:
We cannot pin in `new` because we do not have a location that we know can be pinned.

We could create one, which is option 2 I mention in the other message, but it entails an extra allocation (i.e., makes the abstraction not zero-cost).

Another option is to pass an uninitialised pinned Registration, which is then initialised in new(). But I couldn't find a clean way to do this when the registration is embedded in another struct because the [documentation says](https://doc.rust-lang.org/std/mem/union.MaybeUninit.html#initializing-a-struct-field-by-field) that we cannot initialise individual fields of structs using MaybeUninit.

Reply:
Sorry, I'm clearly thinking slowly, but the implementation right now already involves a heap alloc, is there a way to use this API that doesn't?

Reply:
Sure, we can allocate the registration on the stack, pin it, register, use the device, sleep/wait, then return (the implicit drop would unregister before the memory is freed). The [pin_mut](https://docs.rs/pin-utils/0.1.0/pin_utils/macro.pin_mut.html) macro allows us to do this without any unsafe code.

But this isn't the most common usage. The most common usage would be something like this:
```
struct DeviceContext {
     x: SomeType,
     y: SomeType,
     z: SomeType,
     reg: miscdev::Registration,
}
```

This needs a single allocation. The user would allocate a single block of memory for all fields in the struct then pin the whole thing. To call Registration::register, one would use pinned project (there are crates to do this without any unsafe code, for example [this](https://crates.io/crates/pin-project)).

If we were to allocate inside `Registration::new`, the example DeviceContext above would require two allocations instead.

Reply:
Ok, I'm following now. Can we do `Registration::new_boxed()` and `Registration::new()` then, to let folks who don't mind the extra alloc have a convenient API.

Reply:
Yes, I'll add new_boxed.

Reply:
I actually added `new_pinned` -- since the constructor will have a pinned memory location, I also call register from the it so the returned Registration is ready in one shot on success.